### PR TITLE
REALMC-12113: Add DefaultRule to exported app structure

### PR DESCRIPTION
**Pull request by Amaan95**
## Changes
- Add the Default Rule to the exported app structure for App Structure V1 and V2
- Write and parse the default rule to/from the Data Sources directory for App Structure V2
- Write and parse the default rule to/from the Services directory for App Structure V1

### CODE REVIEW

File: internal/local/app_test.go
Comment (malexandert) [inline internal/local/app_test.go:723]: [q] do we need to specify the `"name"` field here. Technically rule names should only be internal - externally we specify the `database` and `collection` which will eventually get concatenated into a `name`

File: internal/local/structure_test.go
Comment (malexandert) [inline internal/local/structure_test.go:364]: same comment about `name` here

File: internal/local/structure_v2_test.go
Comment (malexandert) [inline internal/local/structure_v2_test.go:51]: [q] why the switch from `Equal` to `Match`?

File: internal/local/structure_v2_test.go
Comment (malexandert) [inline internal/local/structure_v2_test.go:53]: nice 💯

File: internal/local/testdata/full_project/services/mdbSvc/rules/foo.bar.json
Comment (malexandert) [inline internal/local/testdata/full_project/services/mdbSvc/rules/foo.bar.json:4]: same comment about `name` here

**Review by malexandert — APPROVED** [review]
Mainly questions from me, but this looks good! I'd definitely wait till Jamie or Nick have a look before merging though 👍🏾

File: internal/local/app_test.go
Comment (Amaan95) [reply internal/local/app_test.go:723]: That's what I thought and it's what we do for data sources in v2 of the app structure with v2 rules, but for legacy services with rules and mongo services in the v1 structure it does appear that we use the `"name"` field here: 
https://github.com/10gen/realm-cli/blob/3795c0740d25d61bed4de823807e7351fcd06bbf/internal/local/structure.go#L382

this seems to be intentional but I'll defer to someone with historical context (perhaps @makesitgo) on whether this is how things should be handled for v2 rules in the v1 app structure. if not, it would be a fairly simple change to construct the name for a v2 rule using the database and collection names when writing the rule file.

File: internal/local/structure_test.go
Comment (Amaan95) [reply internal/local/structure_test.go:364]: addressed above

File: internal/local/testdata/full_project/services/mdbSvc/rules/foo.bar.json
Comment (Amaan95) [reply internal/local/testdata/full_project/services/mdbSvc/rules/foo.bar.json:4]: addressed above

File: internal/local/structure_v2_test.go
Comment (Amaan95) [reply internal/local/structure_v2_test.go:51]: I had encountered a type match error earlier when testing but that seems to be resolved, will change it back!

File: internal/local/app_test.go
Comment (makesitgo) [reply internal/local/app_test.go:723]: we actually don't seem to ever export `"name"` as a field in any exported rule file for any config version.  see:

* `20200603` and below: https://github.com/10gen/baas/blob/master/importexport/mongodbrules/v2/rule.go#L232-L264
* `20210101` and above: https://github.com/10gen/baas/blob/master/importexport/mongodbrules/v2/rule.go#L312-L328

I think what you're seeing in `local/structure` for the `writeServices` codepath is actually a bug, and we should instead being computing "name" via "database"+"collection" fields.

fortunately (and unfortunately), that codepath is actually never hit in the CLI... I don't love that we needed to go down the route of teaching the CLI how to write _any_ app config files b/c for 98% of the scenarios, the "writing" occurs just by reading the zip response returned from the API.  for context, we went this route so that `app init`/`app create` could link clusters/data sources

File: internal/local/structure_test.go
Comment (makesitgo) [inline internal/local/structure_test.go:364]: would bump the need to remove this (and other instances of it) so we more accurately reflect how the server exports these payloads

File: internal/local/structure_v2.go
Comment (makesitgo) [inline internal/local/structure_v2.go:48]: [question] I'm guessing it's currently a moot point, but perhaps we want to add `DefaultRule` to `HTTPServiceStructure` and its corresponding parse/write methods as well?

basically, in switch from structure v1 => v2, we split "services" into "services+http services+data sources"

if "default rule" is a concept on our core `ServiceDesc` interface in the backend, we may just want to future-proof ourselves and match that design by including it here as well.  it seems that `master` on `baas` is already equipped to handle the same based on your recent changes (b/c data sources and http services all just embed the core `Service` type)

File: internal/local/structure_v2_test.go
Comment (makesitgo) [inline internal/local/structure_v2_test.go:47]: [style/meganit] :meh: ... doubtful this really buys all that much at the expense of one extra hop in `git blame`/`git history`

**Review by makesitgo — COMMENTED** [review]
I wouldn't mind your help with a drive-by bug fix on using `"name"` here plus a few other minor things

File: internal/local/app_test.go
Comment (Amaan95) [reply internal/local/app_test.go:723]: Thanks for the context! I'll make the change 👍🏽

File: internal/local/structure_v2_test.go
Comment (Amaan95) [reply internal/local/structure_v2_test.go:47]: Fair enough 😅

File: internal/local/structure_v2.go
Comment (Amaan95) [reply internal/local/structure_v2.go:48]: Yeah, I don't think there's any value in doing this at the moment given that for all non-mongo services we return a nil value for the default rule and I don't think that's liable to change. 

It's unfortunate that the DefaultRule is a concept on the core `ServiceDesc` but that's a tradeoff we had to make for ease of implementation. Once 3rd party services are deprecated in the future, we may be able to revisit this and have separate implementations for Mongo services and have the DefaultRule be exclusive to that interface.

File: internal/local/structure.go
Comment (makesitgo) [inline internal/local/structure.go:412]: is this actually true?  I'm doubting myself now but I thought I saw in `importexport` that we never exported the "name" field...

**Review by makesitgo — COMMENTED** [review]
I think we should 100% verify if v1 rules (which is a misleading way to refer to them as, since I think you're saying "v1" in an app structure sense but we also commonly refer to these rules as "v2" rules from the mongodb sense) actually export the name field

File: internal/local/structure.go
Comment (makesitgo) [inline internal/local/structure.go:358]: what gets written here if `DefaultRule` is nil or empty?  do we want the file to exist but be empty/`{}`?

I would have guessed maybe we want a `if len(svc.DefaultRule) != 0` check wrapping this file write

File: internal/local/structure.go
Comment (makesitgo) [inline internal/local/structure.go:416]: [nit] always good to keep a tight scope
```suggestion
			if _, isMongoSvcRule := rule["database"]; !isMongoSvcRule {
```

File: internal/local/structure.go
Comment (makesitgo) [inline internal/local/structure.go:418]: we should also always avoid the panic if possible
```suggestion
				if name, ok := rule["name"].(string); ok {
				        ruleName = name
				}
```

**Review by makesitgo — APPROVED** [review]
last few things from me... approving since they're all related to the non-critical "write services" portion of this codebase

File: internal/local/structure.go
Comment (Amaan95) [reply internal/local/structure.go]: Good catch! Added this check as well as a test case to ensure this is handled properly

**Comment by Amaan95** [comment]
CLI tests appear to be failing on master. Failing tests are unrelated to the changes in this PR. Will go ahead and merge.

Failing tests ticket: https://jira.mongodb.org/browse/REALMC-13220

**Pull request merged**
